src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11769,54 +11899,53 @@ internal ExpressionSyntax ParseConsequenceSyntax()
-                {
-                    while (this.CurrentToken.Kind == SyntaxKind.ExclamationToken)
-                        expr = _syntaxFactory.PostfixUnaryExpression(SyntaxKind.SuppressNullableWarningExpression, expr, EatToken());
-                }
+                using var beforeSuppressionsResetPoint = GetDisposableResetPoint(resetOnDispose: false);
+
+                // consume any '!'s from the preceding conditional operation.
+                while (this.CurrentToken.Kind == SyntaxKind.ExclamationToken)
+                    expr = _syntaxFactory.PostfixUnaryExpression(SyntaxKind.SuppressNullableWarningExpression, expr, EatToken());
 
                 switch (this.CurrentToken.Kind)
                 {
+                    // a?.b()
                     case SyntaxKind.OpenParenToken:
-                        expr = _syntaxFactory.InvocationExpression(expr, this.ParseParenthesizedArgumentList());
-                        continue;
+                        return _syntaxFactory.InvocationExpression(expr, this.ParseParenthesizedArgumentList());
 
+                    // a?.b[]
                     case SyntaxKind.OpenBracketToken:
-                        expr = _syntaxFactory.ElementAccessExpression(expr, this.ParseBracketedArgumentList());
-                        continue;
+                        return _syntaxFactory.ElementAccessExpression(expr, this.ParseBracketedArgumentList());
 
+                    // a?.b.
                     case SyntaxKind.DotToken:
-                        expr = _syntaxFactory.MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression, expr, this.EatToken(), this.ParseSimpleName(NameOptions.InExpression));
-                        continue;
+                        return _syntaxFactory.MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression, expr, this.EatToken(), this.ParseSimpleName(NameOptions.InExpression));
 
+                    // a?.b?
                     case SyntaxKind.QuestionToken:
                         return !CanStartConsequenceExpression()
                             ? expr
                             : _syntaxFactory.ConditionalAccessExpression(
                                 expr,
                                 operatorToken: this.EatToken(),
                                 ParseConsequenceSyntax());
-
-                    default:
-                        return expr;
                 }
-            }
 
-            bool isOptionalExclamationsFollowedByConditionalOperation()
-            {
-                var index = 0;
-                while (this.PeekToken(index).Kind == SyntaxKind.ExclamationToken)
-                    index++;
+                var (operatorTokenKind, operatorExpressionKind) = GetOperatorTokenAndExpressionKind();
+                if (IsExpectedAssignmentOperator(operatorTokenKind))
+                {
+                    return ParseAssignmentOperator(operatorExpressionKind, expr, EatOperatorToken(operatorTokenKind));
```

Comment:
I did not think it was possible to further reuse elements of 'ParseExpressionContinued' here but would be happy to be proven wrong.